### DISCUSSION TITLE
MCLK output on Pi Pico/RP2040

### DISCUSSION THREAD

**Discussion by thonk-luke** [pschatzmann/arduino-audio-tools#1029] [Q&A] [answered]
Hey all, I'm messing around with a Pi Pico and an I2S codec (PCM3060) that requires an MCLK and haven't had any luck yet, I've read all the wikis/discussions/examples but can't seem to find a straight answer to if this library supports MCLK output with the Pico or if I'm just wasting my time. Has anybody been able to get a CODEC that requires an MCLK working with the Pico or should I try something else? 
Thanks!

**Comment by pschatzmann** [comment]
Are you asking for the official MBED implementation or for the one from Earle Phil Hower ?

**Reply by thonk-luke** [reply]
Thanks for the reply, I've been using the one from Earle

**Reply by pschatzmann** [reply]
I just committed a [correction](https://github.com/pschatzmann/arduino-audio-tools/commit/a3e546e05e0a7a6c9b7141b9a1e8b54ac1ffe925) to add support for the MCLK

**Reply by thonk-luke** [reply]
Oh wow thank you! So I should just be able to assign a pin using 
`  config.pin_mck = 
` 
and it should work?

**Comment by pschatzmann** [comment]
Sure, under the condition that you upgrade to the latest verison.

The I2S class provided by Hower extends from Stream as well, you could also use that one as output class...

**Reply by thonk-luke** [reply]
Okay I've updated and the code is compiling and uploading but I'm not getting any output on the mck pin. Rest of the clocks are all outputting and measuring fine, are there some conditions on which pin the mck can be on?

```
#include "AudioTools.h"

AudioInfo info(44100, 2, 16);
SineWaveGenerator<int16_t> sineWave(32000);     // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);  // Stream generated from sine wave
I2SStream out;
StreamCopy copier(out, sound);  // copies sound into i2s

// Arduino Setup
void setup(void) {
  // Open Serial
  Serial.begin(115200);
  while (!Serial)
    ;
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.copyFrom(info);
  config.is_master = true;
  config.pin_ws = 21;
  config.pin_bck = 20;
  config.pin_data = 19;
  config.pin_mck = 18;
  out.begin(config);

  // Setup sine wave
  sineWave.begin(info, N_B4);
  Serial.println("started...");
}

// Arduino loop - copy sound to out
void loop() {
  copier.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
I should have read the documentation: https://arduino-pico.readthedocs.io/en/latest/i2s.html

Here is another [correction](https://github.com/pschatzmann/arduino-audio-tools/commit/e9d6580db3f474d24df70d65ea6f79254b62e660)...

It uses cfg.mck_multiplier  = 64; by default. You can change it to another multiple of 64...

**Reply by thonk-luke** [reply]
Amazing, seems to be working great, finally have I2S audio out of my Pico after weeks of pulling my hair out! Thanks so much for your help on this and the quick responses, and the great work you have done!
